core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -659,18 +657,20 @@ public Node start() throws NodeValidationException {
         TransportService transportService = injector.getInstance(TransportService.class);
         transportService.getTaskManager().setTaskResultsService(injector.getInstance(TaskResultsService.class));
         transportService.start();
+        assert localNodeFactory.getNode() != null;
+        assert transportService.getLocalNode().equals(localNodeFactory.getNode())
+            : "transportService has a different local node than the factory provided";
         validateNodeBeforeAcceptingRequests(settings, transportService.boundAddress(), pluginsService.filterPlugins(Plugin.class).stream()
             .flatMap(p -> p.getBootstrapChecks().stream()).collect(Collectors.toList()));
 
         clusterService.addStateApplier(transportService.getTaskManager());
+        clusterService.getMasterService().setClusterStateSupplier(discovery::clusterState);
```

Comment:
We discussed this: Discovery is in charge of the cluster state so that it can make changes to it without going through MasterService (e.g. when a node is a follower) and a future discovery implementation will also be in charge of persisting the cluster state (right now this is done by GatewayMetaState, which is a ClusterStateApplier).